daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/RunnerMain.scala

```diff
@@ -22,122 +22,139 @@ import com.daml.lf.iface.reader.InterfaceReader
-            )
-          )
-      }
-      case None =>
-        Participants(
-          default_participant = Some(
-            ApiParameters(
-              config.ledgerHost.get,
-              config.ledgerPort.get,
-              token,
-              config.applicationId,
-            )
-          ),
-          participants = Map.empty,
-          party_participants = Map.empty,
-        )
-    }
     val flow: Future[Unit] = for {
-
+      config <- Future.fromTry(VerifiedRunnerConfig(runnerConfig))
       clients <-
         if (config.jsonApi) {
-          val ifaceDar = dar.map(pkg => InterfaceReader.readInterface(() => \/-(pkg))._2)
+          val ifaceDar = config.dar.map(pkg => InterfaceReader.readInterface(() => \/-(pkg))._2)
           val envIface = EnvironmentInterface.fromReaderInterfaces(ifaceDar)
           // TODO (#13973) resolve envIface, or not, depending on whether inherited choices are needed
-          Runner.jsonClients(participantParams, envIface)
+          Runner.jsonClients(config.participantParams, envIface)
         } else {
-          Runner.connect(participantParams, config.tlsConfig, config.maxInboundMessageSize)
+          Runner.connect(config.participantParams, config.tlsConfig, config.maxInboundMessageSize)
         }
-      adminClient = LedgerClient.singleHost(
-        hostIp = config.ledgerHost.get,
-        port = config.ledgerPort.get,
-        configuration = LedgerClientConfiguration(
-          applicationId = "admin-client",
-          ledgerIdRequirement = LedgerIdRequirement.none,
-          commandClient = CommandClientConfiguration.default,
-          token = token,
-        ),
-        channelConfig = LedgerClientChannelConfiguration(None),
-      )
-      _ <- adminClient.packageManagementClient.uploadDarFile(
-        ByteString.copyFrom(Files.readAllBytes(config.darPath.toPath))
-      )
-      result <- Runner.run(dar, scriptId, inputValue, clients, config.timeMode)
+      result <- Runner.run(config.dar, config.scriptId, config.inputValue, clients, config.timeMode)
       _ <- Future {
         config.outputFile.foreach { outputFile =>
           val jsVal = LfValueCodec.apiValueToJsValue(result.toUnnormalizedValue)
-          val outDir = outputFile.getParentFile()
+          val outDir = outputFile.getParentFile
           if (outDir != null) {
-            val _ = Files.createDirectories(outDir.toPath())
+            val _ = Files.createDirectories(outDir.toPath)
           }
           Files.write(outputFile.toPath, Seq(jsVal.prettyPrint).asJava)
         }
       }
     } yield ()
 
     flow.onComplete(_ =>
-      if (config.jsonApi) {
+      if (runnerConfig.jsonApi) {
         Http().shutdownAllConnectionPools().flatMap { case () => system.terminate() }
       } else {
         system.terminate()
       }
     )
     Await.result(flow, Duration.Inf)
   }
+
+  final case class VerifiedRunnerConfig private (
```

Comment:
Using a `private` constructor and a `NoCopy` trait we allow the compiler to control who may create or modify our configuration objects after they have been constructed.

Reply:
I think a bunch of other classes use `CliConfig` & `Config` to distinguish between the one set on the Cli and the validated one. You could consider doing something similar.